docs/redux.md

```diff
@@ -0,0 +1,1228 @@
+Emphasize that reducers should be pure and avoid side effects.
+
+```typescript
+// Bad: Performing expensive calculations inside the reducer
+const initialState = { data: [], expensiveResult: 0 };
+
+function myReducer(state = initialState, action) {
+switch (action.type) {
+    case 'ADD_DATA':
+    const newData = action.payload;
+    const expensiveResult = newData.reduce((acc, item) => acc + item.value, 0); // Expensive operation
+    return {
+        ...state,
+        data: [...state.data, newData],
+        expensiveResult,
+    };
+    default:
+        return state;
+    }
+}
+
+// Good: Perform expensive calculations outside the reducer
+const initialState = { data: [], expensiveResult: 0 };
+
+function myReducer(state = initialState, action) {
+    switch (action.type) {
+        case 'ADD_DATA':
+        return {
+            ...state,
+            data: [...state.data, action.payload],
+        };
+        case 'SET_EXPENSIVE_RESULT':
+        return {
+            ...state,
+            expensiveResult: action.payload,
+        };
+        default:
+            return state;
+    }
+}
+
+// Perform the expensive calculation in an action creator or middleware
+function addData(newData) {
+    return (dispatch, getState) => {
+        dispatch({ type: 'ADD_DATA', payload: newData });
+        const expensiveResult = newData.reduce((acc, item) => acc + item.value, 0);
+        dispatch({ type: 'SET_EXPENSIVE_RESULT', payload: expensiveResult });
+    };
+}
+```
+
+### **2.2 Memoization**
+Use selectors and reselect to memoize derived state.
+```typescript
+     import { createSelector } from 'reselect';
+    // State shape
+    const state = {
+    items: [
+        { id: 1, value: 10 },
+        { id: 2, value: 20 },
+    ],
+    };
+
+    // Basic selector
+    const selectItems = (state) => state.items;
+
+    // Memoized selector using reselect
+    const selectTotalValue = createSelector(
+        [selectItems],
+        (items) => items.reduce((total, item) => total + item.value, 0)
+    );
+
+    // Usage
+    const totalValue = selectTotalValue(state); // 30
+```
+### **2.3 Normalization**
+Normalize state shape to avoid deeply nested structures
+```typescript
+    // Bad: Deeply nested state shape
+    const state = {
+    users: {
+        byId: {
+        1: { id: 1, name: 'Alice', posts: [{ id: 1, title: 'Post 1' }] },
+        2: { id: 2, name: 'Bob', posts: [{ id: 2, title: 'Post 2' }] },
+        },
+    },
+    };
+
+    // Good: Normalized state shape
+    const normalizedState = {
+        users: {
+            byId: {
+                1: { id: 1, name: 'Alice', postIds: [1] },
+                2: { id: 2, name: 'Bob', postIds: [2] },
+            },
+            allIds: [1, 2],
+        },
+        posts: {
+            byId: {
+                1: { id: 1, title: 'Post 1' },
+                2: { id: 2, title: 'Post 2' },
+            },
+            allIds: [1, 2],
+        },
+    };
```

Comment:
C: it may not be obvious that these are IDs as they look like indexes. If it's just indexes we could use an array instead, right?
So maybe replace 1, 2 by things that look more like ids like short random hex string for the example to be clear?